### PR TITLE
fix(fees): consolidate Core and Platform fee estimation

### PR DESCRIPTION
**Pull request by lklimek**
## Summary

- **Fix Core (L1) tx fee estimation**: Replace hardcoded `DEFAULT_TX_FEE` of 1000 duffs with size-based fee calculation. Fixes "min relay fee not met" errors for larger transactions (e.g., multi-input asset locks). Deduplicates 3 copies of `estimate_p2pkh_tx_size` into a single utility in `fee_estimation.rs`.
- **Consolidate Platform (L2) fee estimation**: Add unified `PlatformFeeEstimator` methods that compute `max(legacy, transition-based)` fees internally, so callers don't need dual-estimation logic. Reduces public API surface by making internal helpers private and removing dead code.
- **UTXO removal consistency**: `FundWithUtxo` paths in identity registration/top-up now use `remove_selected_utxos()` instead of manual `retain()`.
- **Add comprehensive tests**: 9 tests covering the unified transition-based fee methods plus a regression test for the relay fee fix.

## Test plan

- Automated: `cargo test --all-features --workspace` — includes new fee estimation tests
- Manual test scenarios: [Fee consolidation](docs/ai-design/2026-02-24-fee-consolidation/manual-test-scenarios.md)

🤖 Generated with [Claude Code](https://claude.com/claude-code)

<sub>🤖 Co-authored by [Claudius the Magnificent](https://github.com/lklimek/claudius) AI Agent</sub>

## Summary by CodeRabbit

* **New Features**
  * Unified, iterative fee estimation across wallet, asset-lock and platform flows for more accurate fees and multi-recipient fee subtraction.

* **Bug Fixes**
  * More reliable fee reconciliation with automatic UTXO reloads and improved dust/change handling.
  * Better error propagation in send flows so failures are surfaced and handled gracefully.

* **Documentation**
  * Added comprehensive test scenarios and guidance for fee estimation, asset-lock, platform transfers, and UTXO handling.

### CODE REVIEW

**Comment by lklimek** [comment]
## 🔍 Audit Summary

**Reviewed by:** Claude Code with a 3-agent team:
- `security-auditor` (security-engineer) — fee calculation exploits, overflow risks, UTXO safety, concurrency
- `code-reviewer` (code-reviewer) — cross-artifact consistency, convention adherence, dead code, API design
- `rust-reviewer` (rust-developer) — Rust idioms, type safety, clippy compliance, test quality

### Overall Assessment

Solid consolidation PR. The UTXO safety refactoring (select → build → sign → remove) is a significant improvement over the old eager-removal pattern. Fee estimation deduplication is clean and well-tested. **No critical or high-severity issues found.**

### Findings Summary

| ID | Severity | Category | Location | Description |
|---|---|---|---|---|
| **FEE-001** | MEDIUM | Insecure Design (A04) | `core/mod.rs:407-420` | SPV mode regression: `reload_utxos` is a no-op, but first build already removed UTXOs |
| **FEE-002** | MEDIUM | Code Quality | `fee_estimation.rs:119-121` | Raw byte count used as fee (duffs) instead of `calculate_relay_fee()` |
| **FEE-003** | LOW | Consistency (A04) | `register_identity.rs`, `top_up_identity.rs` | `FundWithUtxo` paths use old `retain()` pattern instead of `remove_selected_utxos()` |
| **FEE-004** | LOW | Code Quality | `send_screen.rs:53-57` | `Result` wrapping on infallible fee estimation adds ceremony |
| **FEE-005** | LOW | Robustness (A05) | `core/mod.rs:429` | `expect()` on Core client lock — would panic on poisoned lock |

### Pre-existing / Outside-diff Issues

These are in code modified by earlier commits already on `v1.0-dev` (not in this PR's diff):

| ID | Severity | Location | Description |
|---|---|---|---|
| **PRE-001** | LOW | `utxos.rs:43` | `tx_out.value as i64` cast is unchecked (safe for Dash supply, but inconsistent with `i64::try_from` used for target) |
| **PRE-002** | LOW | `utxos.rs:64-65` | `total_input as u64 - amount - fee` relies on implicit sign invariants; `checked_sub` would be defense-in-depth |
| **PRE-003** | INFO | `fee_estimation.rs:394-401` | `estimate_fee_from_transition` returns `unwrap_or(0)` — silent failure masked by `max(legacy, transition)` |

### Redundancy Report

- **3/3 agents** flagged FEE-001 (SPV reload regression) — highest redundancy, highest confidence
- **2/3 agents** flagged FEE-002 (raw byte count as fee)
- **2/3 agents** flagged FEE-004 (Result wrapping ceremony)
- **2/3 agents** flagged PRE-001/PRE-002 (UTXO arithmetic safety)

### Positive Observations

1. **Atomic UTXO removal** — `select_unspent_utxos_for` (non-mutating) + `remove_selected_utxos` (after success) prevents fund loss on tx build failure
2. **Thorough test coverage** — 14 Core fee tests + 9 unified Platform fee tests, including edge cases (dust absorption, change elimination, overflow)
3. **Clean deduplication** — 3 copies of `estimate_p2pkh_tx_size` → 1 canonical location
4. **Good API hygiene** — internal helpers made private, dead code removed, `AssetLockFeeResult` struct replaces ad-hoc tuples
5. **DB-before-broadcast** — asset lock stored before broadcast prevents race with fast InstantSend
6. **Lock poisoning fixes** — several `unwrap()` calls replaced with `map_err()`

<sub>🤖 Co-authored by [Claudius the Magnificent](https://github.com/lklimek/claudius) AI Agent</sub>

File: src/backend_task/core/mod.rs [hunk 1]
```diff
@@ -372,22 +372,55 @@ impl AppContext {
...
+            // If override_fee was supplied, trust it — skip recalculation.
+            if request.override_fee.is_none() {
+                let actual_fee =
+                    calculate_relay_fee(estimate_p2pkh_tx_size(tx.input.len(), tx.output.len()));
+                if actual_fee > initial_fee {
+                    // The real fee is higher; rebuild with the correct fee.
+                    // The first build already removed UTXOs, so reload the wallet
+                    // state and rebuild.  In practice this rarely triggers because
+                    // the initial 5-input estimate is usually sufficient.
+                    wallet_guard.reload_utxos(self)?;
```
Comment (lklimek) [inline src/backend_task/core/mod.rs]: **FEE-001 (MEDIUM)** — SPV mode regression (A04: Insecure Design)

`reload_utxos()` is a no-op in SPV mode (returns `Ok(false)`), but the first `build_multi_recipient_payment_transaction` on line 395 already removed UTXOs via `remove_selected_utxos`. If fee recalculation triggers in SPV mode, the second build operates on a depleted UTXO set → spurious "Insufficient funds" error.

This function is named `_via_rpc` so it's presumably RPC-only, but there's no guard.

**Recommendation:** Add a guard on the `reload_utxos` return value:
```rust
if !wallet_guard.reload_utxos(self)? {
    return Err("Fee re-estimation failed: cannot reload UTXOs in SPV mode".to_string());
}
```
Or restructure to not commit UTXO removal in the first build attempt.

File: src/model/fee_estimation.rs [hunk 2]
```diff
@@ -12,7 +21,163 @@
...
+pub fn calculate_asset_lock_fee(
+    total_input_value: u64,
+    requested_amount: u64,
+    num_inputs: usize,
+    allow_take_fee_from_amount: bool,
+) -> Result<AssetLockFeeResult, String> {
+    // First pass: assume 2 outputs (1 burn + 1 change).
+    let fee_with_change = std::cmp::max(
+        MIN_ASSET_LOCK_FEE,
+        estimate_asset_lock_tx_size(num_inputs, 2) as u64,
```
Comment (lklimek) [inline src/model/fee_estimation.rs]: **FEE-002 (MEDIUM)** — Raw byte count used as fee amount

`estimate_asset_lock_tx_size()` returns a **byte count**, but it's compared directly against `MIN_ASSET_LOCK_FEE` (3000 **duffs**). This works because Dash's relay fee is ~1 duff/byte, but `calculate_relay_fee()` exists and is used for P2PKH fees.

For a 21-input tx: raw size = 3248 bytes → used as 3248 duffs. With `calculate_relay_fee(3248)` the result differs due to rounding.

**Recommendation:** Use `calculate_relay_fee(estimate_asset_lock_tx_size(num_inputs, 2))` for consistency, or add a doc comment explaining why asset lock fees intentionally use size-as-fee.

File: src/backend_task/core/mod.rs [hunk 3]
```diff
@@ -372,28 +372,79 @@ impl AppContext {
...
         let parsed_recipients = self.parse_recipients(&request)?;
 
-        const DEFAULT_TX_FEE: u64 = 1_000;
+        // Iterative fee estimation: estimate fee based on assumed input count,
+        // build the transaction, then verify the fee covers the actual tx size.
+        // If the caller supplied an override_fee (e.g. retry after a min-relay-
+        // fee rejection), use that directly and skip the iteration.
+        let num_recipient_outputs = parsed_recipients.len() + 1; // +1 for change
+        let initial_fee = request.override_fee.unwrap_or_else(|| {
+            calculate_relay_fee(estimate_p2pkh_tx_size(5, num_recipient_outputs))
```
Comment (PastaPastaPasta) [inline src/backend_task/core/mod.rs:385]: @coderabbitai does this say that we are estimating based on 5 inputs? I think this would massively overpay fees in typical case of 1->2 or 2->2 payments.

Comment (PastaPastaPasta) [reply src/backend_task/core/mod.rs:385]: @lklimek please fix

**Review by PastaPastaPasta — CHANGES_REQUESTED** [review]
Issues & Concerns

  1. UTXO loss in SPV mode during fee re-estimation (Medium severity)

  In src/backend_task/core/mod.rs, the iterative fee path:

  let mut tx = wallet_guard.build_multi_recipient_payment_transaction(..., initial_fee, ...)?;

  if actual_fee > initial_fee {
      if !wallet_guard.reload_utxos(self)? {
          return Err("Fee re-estimation failed: cannot reload UTXOs in SPV mode".to_string());
      }
      // ... rebuild
  }

  build_multi_recipient_payment_transaction calls remove_selected_utxos() at the end of the first build. If fee re-estimation triggers and
  reload_utxos returns false (SPV mode), the error is returned but the UTXOs from the first build are already removed from the wallet's
  in-memory state. This means in SPV mode, a failed fee re-estimation permanently loses UTXOs until the next full sync.

  Suggestion: Either (a) pre-check whether fee re-estimation would be needed before the first build (compare initial_fee against a worst-case
  estimate), or (b) save the UTXO state before the first build and restore it on the SPV error path. Or (c) add a comment acknowledging this as
  an acceptable risk given the 5-input initial estimate is generous.

  2. Sanity check warning without recovery (Low severity)

  if actual_fee < rebuilt_fee {
      tracing::warn!("Fee still insufficient after rebuild: {} < {}", actual_fee, rebuilt_fee);
  }

  This logs a warning but proceeds to broadcast a transaction with insufficient fee. This could lead to a network rejection. Consider either
  returning an error here or doing a third iteration.

**Comment by thepastaclaw** [comment]
Closing as superseded by the current fee/payment architecture. Current `v1.0-dev` has size-based transaction estimation, min-relay-fee retry handling, and an expanded `PlatformFeeEstimator`; this branch targets older Core/SPV payment paths and also retains unresolved requested-changes findings. Any remaining fee bug should be filed from a current reproduction rather than revived from this implementation.
